File: android-dpc/app/src/test/kotlin/io/github/helios57/familyguard/recovery/RecoveryVectorsTest.kt
package io.github.helios57.familyguard.recovery

import io.github.helios57.familyguard.net.RecoveryMaterial
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.jsonObject
import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.util.Base64

/**
 * Replays `backend/internal/auth/recovery-vectors.json` — the same file, the same cases and the same
 * expectations as the Go side's `TestRecoveryVectors`.
 *
 * This is the only control that can catch the two implementations of the recovery code drifting
 * apart, and drift here is silent on both sides: every test in this repository stays green while the
 * phone rejects the twenty characters printed on the parent's screen. By then the device is locked
 * down, possibly offline, and there is no way to read anything back off it.
 *
 * The file is not generated from either side. Its normalisation expectations are hand-written from
 * the rule in `tokens.go`, and its digests come from Python's `hashlib` — a third implementation of
 * RFC 8018. Vectors taken from Go's output would assert that Go does what Go does; taken from this
 * side's, that Kotlin does what Kotlin does. Neither is evidence that the two agree.
 */
class RecoveryVectorsTest {

    @Serializable
    private data class NormalizeVector(
        @SerialName("name") val name: String,
        @SerialName("raw") val raw: String,
        @SerialName("expect") val expect: String,
    )

    @Serializable
    private data class DeriveVector(
        @SerialName("name") val name: String,
        @SerialName("code") val code: String,
        @SerialName("salt") val salt: String,
        @SerialName("iterations") val iterations: Int,
        @SerialName("expect_hash") val expectHash: String,
    )

    @Serializable
    private data class VerifyVector(
        @SerialName("name") val name: String,
        @SerialName("code") val code: String,
        @SerialName("salt") val salt: String,
        @SerialName("iterations") val iterations: Int,
        @SerialName("hash") val hash: String,
        @SerialName("expect") val expect: Boolean,
    )

    /**
     * Rejects unknown keys, for the reason the Go loader states: a misspelled key would otherwise be
     * dropped in silence and the vector would assert against a default. An `iterations` typo would
     * run every derive case at zero rounds, which fails — and reads as an implementation correctly
     * refusing a zero work factor rather than as a broken vector file.
     */
    private val strict = Json { ignoreUnknownKeys = false }

    private inline fun <reified T> section(name: String): List<T> {
        // Copied onto the test classpath by copyRecoveryVectors, which fails the build if the source
        // has moved. A test reading `../../../backend/...` would survive that refactor by finding
        // nothing and asserting over zero vectors.
        val raw = javaClass.classLoader?.getResourceAsStream("recovery-vectors.json")?.use {
            it.readBytes().decodeToString()
        }
        if (raw.isNullOrEmpty()) {
            throw AssertionError(
                "recovery-vectors.json is not on the test classpath. It is copied there from " +
                    "backend/internal/auth/ by the copyRecoveryVectors task; without it this suite " +
                    "asserts nothing and must not be read as a pass.",
            )
        }
        val array = strict.parseToJsonElement(raw).jsonObject[name] as? JsonArray
            ?: throw AssertionError("the shared recovery vectors have no \"$name\" section")
        assertTrue("the \"$name\" section is empty, so replaying it asserts nothing", array.isNotEmpty())
        return array.map { strict.decodeFromJsonElement<T>(it) }
    }

    /** RawURL, matching `base64.RawURLEncoding` on the server: the decoder tolerates the absent padding. */
    private fun decode(value: String): ByteArray =
        if (value.isEmpty()) ByteArray(0) else Base64.getUrlDecoder().decode(value)

    @Test
    fun `every normalize vector folds the way the server folds it`() {
        val seen = mutableSetOf<String>()
        for (v in section<NormalizeVector>("normalize")) {
            assertTrue("duplicate normalize vector \"${v.name}\"", seen.add(v.name))
            assertEquals(
                "vector \"${v.name}\": normalize(${v.raw.escaped()})",
                v.expect,
                RecoveryCode.normalize(v.raw),
            )
        }
        requirePresent(seen, DIVERGENCE_VECTORS, "normalize")
    }

    @Test
    fun `every derive vector produces the digest a third implementation produced`() {
        val seen = mutableSetOf<String>()
        for (v in section<DeriveVector>("derive")) {
            assertTrue("duplicate derive vector \"${v.name}\"", seen.add(v.name))
            val want = decode(v.expectHash)
            assertEquals("vector \"${v.name}\": digest length", KEY_BYTES, want.size)

            // normalize-then-derive, which is what DeriveRecoveryHash does on the other side in one
            // call. Split here because RecoveryDerivation takes the folded code: the verifier has to
            // be testable without a JCE provider, so the fold happens outside it.
            val got = Pbkdf2HmacSha256.derive(RecoveryCode.normalize(v.code), decode(v.salt), v.iterations)
            assertArrayEquals("vector \"${v.name}\": digest", want, got)
        }
        assertTrue(
            "no derive vector runs at the production work factor, so nothing here checks the number " +
                "a device actually runs",
            seen.contains(PRODUCTION_WORK_FACTOR_VECTOR),
        )
    }

    @Test
    fun `every verify vector is accepted or refused the way the server decides`() {
        val seen = mutableSetOf<String>()
        var accept = 0
        var reject = 0
        for (v in section<VerifyVector>("verify")) {
            assertTrue("duplicate verify vector \"${v.name}\"", seen.add(v.name))
            if (v.expect) accept++ else reject++

            val verifier = RecoveryVerifier(
                RecoveryMaterial(salt = v.salt, iterations = v.iterations, hash = v.hash),
            )
            assertEquals("vector \"${v.name}\"", v.expect, verifier.verify(v.code))
        }

        // A section of only accepting cases is passed by a verifier that returns true, and one of
        // only rejecting cases by a verifier that returns false. Both are what a half-finished
        // mirror looks like.
        assertTrue(
            "verify vectors: $accept accepting, $reject rejecting — a section that is all one " +
                "answer is passed by a constant",
            accept > 0 && reject > 0,
        )
    }

    /**
     * The vectors that exist *because* the two platforms differ, pinned by name.
     *
     * Not a count. The failure this guards against is not a truncated file — that fails to parse —
     * but the plausible, well-meant edit: a mirror fails one of these, and the vector is deleted or
     * softened to make the build green. A count would be updated in the same breath. A missing name
     * says which specific disagreement stopped being checked.
     */
    private fun requirePresent(seen: Set<String>, required: List<String>, section: String) {
        val missing = required.filterNot(seen::contains)
        assertTrue(
            "the $section section no longer contains ${missing.joinToString(", ") { "\"$it\"" }}. " +
                "Each of these pins a known Go/Java divergence, so removing one does not make the " +
                "two sides agree — it stops measuring whether they do.",
            missing.isEmpty(),
        )
    }

    /** Renders control characters and blanks visibly, so a failure message can be read. */
    private fun String.escaped(): String =
        "\"" + map { c ->
            when {
                c == '"' -> "\\\""
                c == '\\' -> "\\\\"
                c.code in 0x20..0x7E -> c.toString()
                else -> "\\u%04x".format(c.code)
            }
        }.joinToString("") + "\""

    private companion object {
        const val KEY_BYTES = 32

        /**
         * U+0085, U+00A0, U+2007 and U+202F are whitespace to Go and not to `Character.isWhitespace`;
         * U+001C..U+001F are the reverse; and `String.uppercase()` turns ß into SS where Go's simple
         * mapping leaves it. Each of these folds differently on the two sides unless the mirror is
         * written deliberately — which is the whole reason this suite exists.
         *
         * Those four are the *complete* Go-minus-Java set, so a mirror that passes this list has no
         * remaining whitespace disagreement to find in production.
         */
        val DIVERGENCE_VECTORS = listOf(
            "a non-breaking space at each end and in the middle, as a chat app pastes it",
            "the four spaces Go strips that Java does not call whitespace at all",
            "an ASCII file separator survives, because Go does not call it a space",
            "a sharp s stays one character, because both sides map case simply",
        )

        const val PRODUCTION_WORK_FACTOR_VECTOR =
            "the production work factor, so the number that ships is the number that is checked"
    }
}

File: backend/internal/auth/recovery_vectors_test.go
package auth

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"testing"
	"unicode"
)

type normalizeVector struct {
	Name   string `json:"name"`
	Raw    string `json:"raw"`
	Expect string `json:"expect"`
}

type deriveVector struct {
	Name       string `json:"name"`
	Code       string `json:"code"`
	Salt       string `json:"salt"`
	Iterations int    `json:"iterations"`
	ExpectHash string `json:"expect_hash"`
}

type verifyVector struct {
	Name       string `json:"name"`
	Code       string `json:"code"`
	Salt       string `json:"salt"`
	Iterations int    `json:"iterations"`
	Hash       string `json:"hash"`
	Expect     bool   `json:"expect"`
}

// decodeVectors pulls one section out of the shared file.
//
// DisallowUnknownFields throughout: a misspelled key would otherwise be dropped in silence and the
// vector would assert against a zero value instead of the one that was written down — an
// `iterations` typo would make every derive case run at zero rounds and refuse, which reads as a
// correctly failing implementation.
func decodeVectors[T any](t *testing.T, section string) []T {
	t.Helper()
	var top map[string]json.RawMessage
	if err := json.Unmarshal(RecoveryVectors, &top); err != nil {
		t.Fatalf("shared recovery vectors are not valid JSON: %v", err)
	}
	raw, ok := top[section]
	if !ok {
		t.Fatalf("shared recovery vectors have no %q section", section)
	}
	var items []json.RawMessage
	if err := json.Unmarshal(raw, &items); err != nil {
		t.Fatalf("%q is not an array: %v", section, err)
	}
	if len(items) == 0 {
		t.Fatalf("the %q section is empty, so replaying it asserts nothing", section)
	}
	out := make([]T, 0, len(items))
	for i, item := range items {
		dec := json.NewDecoder(bytes.NewReader(item))
		dec.DisallowUnknownFields()
		var v T
		if err := dec.Decode(&v); err != nil {
			t.Fatalf("%s vector %d is malformed: %v", section, i, err)
		}
		out = append(out, v)
	}
	return out
}

func mustDecodeB64(t *testing.T, value string) []byte {
	t.Helper()
	if value == "" {
		return nil
	}
	// RawURLEncoding, matching what the API emits and what the DPC decodes. Padded input is a
	// vector that was written by hand against the wrong encoder, so it fails here rather than
	// silently producing different bytes on the two sides.
	decoded, err := base64.RawURLEncoding.DecodeString(value)
	if err != nil {
		t.Fatalf("vector value %q is not base64 RawURL: %v", value, err)
	}
	return decoded
}

// TestRecoveryVectors replays the shared file. It is the anti-drift control for FR-12.3: the same
// vectors run in the DPC's JVM suite, so a normaliser or a work factor the two sides disagree about
// turns a build red instead of producing a phone that rejects the code the parent was shown.
//
// Note what the file is NOT: it is not generated from this package. The normalisation expectations
// are hand-written from the rule in NormalizeRecoveryCode's doc comment and the digests come from
// Python's hashlib — a third PBKDF2 implementation. Regenerating it from Go output would leave it
// asserting only that Go does what Go does.
func TestRecoveryVectors(t *testing.T) {
	t.Run("normalize", func(t *testing.T) {
		seen := map[string]bool{}
		for _, v := range decodeVectors[normalizeVector](t, "normalize") {
			if v.Name == "" {
				t.Fatal("a normalize vector has no name")
			}
			if seen[v.Name] {
				t.Fatalf("duplicate normalize vector %q — a rename would silently drop a case", v.Name)
			}
			seen[v.Name] = true
			if got := NormalizeRecoveryCode(v.Raw); got != v.Expect {
				t.Errorf("%s: NormalizeRecoveryCode(%q) = %q, want %q", v.Name, v.Raw, got, v.Expect)
			}
		}
	})

	t.Run("derive", func(t *testing.T) {
		seen := map[string]bool{}
		for _, v := range decodeVectors[deriveVector](t, "derive") {
			if v.Name == "" {
				t.Fatal("a derive vector has no name")
			}
			if seen[v.Name] {
				t.Fatalf("duplicate derive vector %q", v.Name)
			}
			seen[v.Name] = true

			want := mustDecodeB64(t, v.ExpectHash)
			if len(want) != recoveryKeyLen {
				t.Fatalf("%s: expects a %d-byte digest, not %d — the vector and the key length disagree",
					v.Name, recoveryKeyLen, len(want))
			}
			got, err := DeriveRecoveryHash(v.Code, mustDecodeB64(t, v.Salt), v.Iterations)
			if err != nil {
				t.Fatalf("%s: DeriveRecoveryHash: %v", v.Name, err)
			}
			if !bytes.Equal(got, want) {
				t.Errorf("%s: digest is %s, want %s",
					v.Name,
					base64.RawURLEncoding.EncodeToString(got),
					base64.RawURLEncoding.EncodeToString(want))
			}
		}
	})

	t.Run("verify", func(t *testing.T) {
		seen := map[string]bool{}
		for _, v := range decodeVectors[verifyVector](t, "verify") {
			if v.Name == "" {
				t.Fatal("a verify vector has no name")
			}
			if seen[v.Name] {
				t.Fatalf("duplicate verify vector %q", v.Name)
			}
			seen[v.Name] = true

			got := VerifyRecoveryCode(v.Code, mustDecodeB64(t, v.Salt), v.Iterations, mustDecodeB64(t, v.Hash))
			if got != v.Expect {
				t.Errorf("%s: VerifyRecoveryCode = %v, want %v", v.Name, got, v.Expect)
			}
		}
	})

	// A file with only accepting cases would pass against a verifier that returns true, and one with
	// only rejecting cases would pass against a verifier that returns false. Neither is hypothetical:
	// both are what a half-finished mirror looks like.
	t.Run("the file itself discriminates", func(t *testing.T) {
		var accept, reject int
		for _, v := range decodeVectors[verifyVector](t, "verify") {
			if v.Expect {
				accept++
			} else {
				reject++
			}
		}
		if accept == 0 || reject == 0 {
			t.Fatalf("verify vectors: %d accepting, %d rejecting — a section that is all one answer "+
				"is passed by a constant", accept, reject)
		}
	})
}

// TestRecoveryAlphabetIsAscii is what lets the DPC use the JCE's PBKDF2 at all.
//
// The password reaches a JCE provider as a char[], and providers disagree about how they turn one
// into bytes: the JDK's encodes UTF-8, and Bouncy Castle's PKCS#5 v2 scheme has historically taken
// the low byte of each character. Go encodes UTF-8. The two agree for every ASCII character and may
// differ above it — so as long as every character a generated code can contain is ASCII, the
// question cannot arise. This test is the thing that keeps that true: adding a character above
// U+007F to recoveryAlphabet turns it red here, in the repository that owns the alphabet, rather
// than on a phone.
//
// It also asserts the exclusions, because they are a product decision (a code is read aloud) that a
// tidying edit would otherwise undo without any test noticing.
func TestRecoveryAlphabetIsAscii(t *testing.T) {
	if len(recoveryAlphabet) == 0 {
		t.Fatal("recoveryAlphabet is empty")
	}
	seen := map[rune]bool{}
	for _, r := range recoveryAlphabet {
		if r > unicode.MaxASCII {
			t.Fatalf("recoveryAlphabet contains %q (U+%04X), which is not ASCII: the DPC's JCE "+
				"provider and Go may encode it differently, and the phone would then reject a "+
				"code the console accepted", r, r)
		}
		if seen[r] {
			t.Fatalf("recoveryAlphabet contains %q twice, which skews the draw", r)
		}
		seen[r] = true
		if isRecoverySeparator(r) {
			t.Fatalf("recoveryAlphabet contains %q, which normalisation strips — a code containing "+
				"it could never be typed back in", r)
		}
	}
	for _, excluded := range "ILOU01" {
		if seen[excluded] {
			t.Errorf("recoveryAlphabet contains %q; the exclusions exist so a code read aloud does "+
				"not depend on telling O from 0", excluded)
		}
	}
}
